### PR TITLE
wip: start exploring some new ideas (Lipo-suction and EVEN MORE LAZY!)

### PR DESCRIPTION
**Pull request by tjdevries**

### CODE REVIEW

**Comment by dam9000** [comment]
How about making it modular, I think that would be a welcome move by the majority of kickstart users. The single file init.lua is hard to read. In this PR a couple of sections are rearranged and it already makes the diff hard to follow. If the categories would be split into separate files this could be avoided.

**Comment by feoh** [comment]
> No. A major benefit is that you can get kickstart working with a simple copy paste if you want. I will show how to extend with modular examples but I think it's overkill and over-complicates the onboarding for a brand new user.
> 
> I don't think diff history is a good enough tradeoff here

Not arguing for modular versus not here. Zero skin in that game..

BUUUUT you can't just copy/paste init.lua into place, and you haven't been able to since the Lazy refactor. It errors out if you do.

You have to clone the entire repo as discussed in the README.

I'd be super happy if we could go back to making your original video work for people because ZOMG do we get a lot of unhappy campers trying that, failing, and reporting bugs :\

File: init.lua [hunk 1]
```diff
@@ -121,12 +309,87 @@ require('lazy').setup({
...
       'rafamadriz/friendly-snippets',
     },
+    config = function()
+      -- See `:help cmp`
+      local cmp = require 'cmp'
+      local luasnip = require 'luasnip'
+      require('luasnip.loaders.from_vscode').lazy_load()
+      luasnip.config.setup {}
+
+      cmp.setup {
```
Comment (kdheepak) [inline init.lua:319]:
```suggestion
      cmp.setup {
          preselect = cmp.PreselectMode.None,
```

File: lua/kickstart/plugins/autoformat.lua
Comment (feoh) [inline lua/kickstart/plugins/autoformat.lua:1]: I like this change! The autoformat thing has caused a bunch of headaches for people, especially since we put the stylua check in place.

**Comment by feoh** [comment]
> How about making it modular, I think that would be a welcome move by the majority of kickstart users. The single file init.lua is hard to read. In this PR a couple of sections are rearranged and it already makes the diff hard to follow. If the categories would be split into separate files this could be avoided.

I've thought this myself *BUT* are we suffering from the [curse of knowledge](https://en.wikipedia.org/wiki/Curse_of_knowledge) ?

One might argue that when you reach the point where you want to start slicing and dicing kickstart, maybe you should do so in a fork :)

File: init.lua [hunk 2]
```diff
@@ -36,6 +40,9 @@ I hope you enjoy your Neovim journey,
 - TJ
 
 P.S. You can delete this when you're done too. It's your config now :)
+
+TODO: Include PDE video link? ;)
```
Comment (feoh) [inline init.lua:44]: Love this! But we should be sure that simply copying in init.lua still works. Last we checked - it doesn't.

File: init.lua [hunk 3]
```diff
@@ -121,12 +309,87 @@ require('lazy').setup({
...
+        mapping = cmp.mapping.preset.insert {
+          ['<C-n>'] = cmp.mapping.select_next_item(),
+          ['<C-p>'] = cmp.mapping.select_prev_item(),
+          ['<C-b>'] = cmp.mapping.scroll_docs(-4),
+          ['<C-f>'] = cmp.mapping.scroll_docs(4),
+          ['<C-Space>'] = cmp.mapping.complete {},
+          ['<CR>'] = cmp.mapping.confirm {
+            behavior = cmp.ConfirmBehavior.Replace,
+            select = true,
+          },
```
[the remark this answers is not in the document]
Comment (kdheepak) [reply init.lua:337]: Add this line too: `          preselect = cmp.PreselectMode.None,`

**Comment by kdheepak** [comment]
https://github.com/nvim-lua/kickstart.nvim/assets/1813121/b64c11c0-7878-4503-87c0-a7dd401535ad

**Review by feoh — COMMENTED** [review]
Holy cow this is amazing work. Thank you!

**Comment by yordis** [comment]
image

**Comment by feoh** [comment]
> cmp-no-select-enter.mov

OK so if I'm understanding the intent of this video correctly, you're saying that having a default selection impedes the flow of your typing when you don't want what's been selected.

Gotcha.

@tjdevries Do you encounter this and if so how do you work around it?

**Comment by rubiin** [comment]
Might well also add lsp-zero and heirline

**Comment by feoh** [comment]
> Might well also add lsp-zero and heirline

Pardon my ignorance but isn't that what we're using Mason for?

**Comment by rubiin** [comment]
If you mean mason-lspconfig, then yeah

File: init.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (kdheepak) [reply init.lua]: This was part of the discussion to get `Enter` working in a way that was "intuitive", but @tjdevries scrapped using `Enter` in favor of `<c-y>` which is more vim-like. This suggestion can be ignored.

File: init.lua (hunk 1 above)
Comment (kdheepak) [reply init.lua:339]: This suggestion can be ignored as well.

**Comment by echasnovski** [comment]
@tjdevries, here is my <d>humble</s> draft suggestion for including 'mini.nvim':

```lua
  { -- Collection of various small independent plugins/modules
    'echasnovski/mini.nvim',
    config = function()
      -- Better Around/Inside textobjects
      -- (`af` - around function call, `inq` - inside next quote, etc.)
      require('mini.ai').setup({ n_lines = 500 })

      -- Go forward/backward with square brackets
      -- (`[b` - previous buffer, `]c` - next comment, etc.)
      require('mini.bracketed').setup({ n_lines = 500 })

      -- File explorer with manipulation by editing text
      require('mini.files').setup()

      -- More operators: duplicate, exchange, replace text
      require('mini.operators').setup()

      -- Add/delete/replace surroundings (brackets, quotes, etc.)
      require('mini.surround').setup()

      -- ... and there is more
    end,
  },
```

This allows small number of lines with more functionality and call to action to explore more.

**Comment by kdheepak** [comment]
Here's some additional mappings that I think are good defaults:

```lua
-- use `+` and `-` to increment and decrement
vim.keymap.set('n', '+', '<C-a>')
vim.keymap.set('n', '-', '<C-x>')
vim.keymap.set('v', '+', 'g<C-a>gv')
vim.keymap.set('v', '-', 'g<C-x>gv')

-- `U` to undo
vim.keymap.set('n', 'U', '<C-R>')

-- Use `Q` to play the macro recorded in `@q`
vim.keymap.set('n', 'Q', '@q')

-- Better indenting
vim.keymap.set('v', '<', '<gv')
vim.keymap.set('v', '>', '>gv')
vim.keymap.set('n', '>', '>>_')
vim.keymap.set('n', '<', '<<_')

```

**Comment by rubiin** [comment]
Here some good keymaps for moving lines 

```
vim.keymap.set("n", "<A-j>", "<cmd>m .+1<cr>==", { desc = "Move down" })
vim.keymap.set("n", "<A-k>", "<cmd>m .-2<cr>==", { desc = "Move up" })
vim.keymap.set("i", "<A-j>", "<esc><cmd>m .+1<cr>==gi", { desc = "Move down" })
vim.keymap.set("i", "<A-k>", "<esc><cmd>m .-2<cr>==gi", { desc = "Move up" })
vim.keymap.set("v", "<A-j>", ":m '>+1<cr>gv=gv", { desc = "Move down" })
vim.keymap.set("v", "<A-k>", ":m '<-2<cr>gv=gv", { desc = "Move up" })
```

**Comment by feoh** [comment]
@rubiin @dam9000 recommended a plugin that does those move commands and more, but I cannot find his recommendation sadlyh.

@dam9000 would you be so kind as to please repeat yourself here? :) I'll take note of it this time I swear because folks keep asking :)

**Comment by echasnovski** [comment]
> @rubiin @dam9000 recommended a plugin that does those move commands and more, but I cannot find his recommendation sadlyh.

'mini.move' ;)

**Comment by feoh** [comment]
> > @rubiin @dam9000 recommended a plugin that does those move commands and more, but I cannot find his recommendation sadlyh.
> 
> 'mini.move' ;)

OOH that looks purrfect! :)

I for one embrace our new mini.* overlords!

**Comment by laughingman-hass** [comment]
when doing a clean install, tree-sitter doesn't seem to auto install the parser for lua. It raises the following error when entering a lua file like `init.lua`

- clean up state and share `rm -rf ~/.local/share/kickstart/ ~/.local/state/kickstart/`
- start nvim with `NVIM_APPNAME=kickstart nvim` and let lazy.vnim download all the plugins, which it does successfully.
- Then start nvim with `NVIM_APPNAME=kickstart nvim init.lua`

```
Error detected while processing BufReadPost Autocommands for "*":
Error executing lua callback: ...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:30: Error executing lua: ...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/fi
letype.lua:31: BufReadPost Autocommands for "*"..FileType Autocommands for "*"..function <SNR>1_LoadFTPlugin[20]..script /opt/homebrew/Cellar/neovim/HEAD-fbe40ca/sh
are/nvim/runtime/ftplugin/lua.lua: Vim(runtime):E5113: Error while calling lua chunk: ...e40ca/share/nvim/runtime/lua/vim/treesitter/language.lua:103: no parser for
 'lua' language, see :help treesitter-parsers
stack traceback:
        [C]: in function 'error'
        ...e40ca/share/nvim/runtime/lua/vim/treesitter/language.lua:103: in function 'add'
        ...a/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:113: in function 'new'
        ...m/HEAD-fbe40ca/share/nvim/runtime/lua/vim/treesitter.lua:64: in function '_create_parser'
        ...m/HEAD-fbe40ca/share/nvim/runtime/lua/vim/treesitter.lua:131: in function 'get_parser'
        ...m/HEAD-fbe40ca/share/nvim/runtime/lua/vim/treesitter.lua:468: in function 'start'
        .../neovim/HEAD-fbe40ca/share/nvim/runtime/ftplugin/lua.lua:2: in main chunk
        [C]: in function 'nvim_cmd'
        ...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:31: in function <...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:30>
        [C]: in function 'nvim_buf_call'
        ...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:30: in function <...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_cmd'
        ...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:31: in function <...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:30>
        [C]: in function 'nvim_buf_call'
        ...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:30: in function <...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_buf_call'
        ...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:30: in function <...llar/neovim/HEAD-fbe40ca/share/nvim/runtime/filetype.lua:10>
Press ENTER or type command to continue
```

File: init.lua [hunk 4]
```diff
@@ -44,60 +80,435 @@ P.S. You can delete this when you're done too. It's your config now :)
 vim.g.mapleader = ' '
 vim.g.maplocalleader = ' '
 
+-- [[ Setting options ]]
+-- See `:help vim.opt`
+-- NOTE: You can change these options as you wish!
+--  For more options, you can see `:help option-list`
+
+-- Set highlight on search
+vim.opt.hlsearch = false
```
Comment (kdheepak) [inline init.lua:89]: Can you elaborate on why this is a good default? Doesn't this hide useful information when searching and hitting `n` to go to the next search.

Comment (kdheepak) [reply init.lua:89]: I like this combination as a better default:

```lua
-- Set highlight on search
vim.opt.hlsearch = true

-- [[ Highlight on yank ]]
-- See `:help vim.highlight.on_yank()`
vim.api.nvim_create_autocmd('TextYankPost', {
  group = vim.api.nvim_create_augroup('YankHighlight', { clear = true }),
  callback = function()
    vim.highlight.on_yank()
  end,
})

vim.keymap.set('n', '<CR>', ':nohlsearch<CR>', { silent = true })

```

**Comment by kdheepak** [comment]
I like restoring the cursor position to where a user last was when reopening a file:

https://github.com/farmergreg/vim-lastplace

Alternatively a one line version:

```lua
-- Restore cursor position
vim.api.nvim_create_autocmd('BufReadPost', { command = 'silent! normal! g`"zv' })
```

**Comment by kdheepak** [comment]
Potentially a good default?

```lua
--  show 10 lines above and below cursor when scrolling
vim.o.scrolloff = 10 
```

File: init.lua [hunk 5]
```diff
@@ -44,60 +80,435 @@ P.S. You can delete this when you're done too. It's your config now :)
...
+
+-- [[ Basic Keymaps ]]
+
+-- Keymaps for better default experience
+-- See `:help vim.keymap.set()`
+vim.keymap.set({ 'n', 'v' }, '<Space>', '<Nop>', { silent = true })
+
+-- Remap for dealing with word wrap
+vim.keymap.set('n', 'k', "v:count == 0 ? 'gk' : 'k'", { expr = true, silent = true })
+vim.keymap.set('n', 'j', "v:count == 0 ? 'gj' : 'j'", { expr = true, silent = true })
```
Comment (kdheepak) [inline init.lua:137]: What do you think about this in addition to the above? 

```lua
-- make gj and gk scroll without line wrap
-- useful when recording macros
vim.keymap.set('n', 'gk', 'k', { expr = true, silent = true })
vim.keymap.set('n', 'gj', 'j', { expr = true, silent = true })
```

File: init.lua [hunk 6]
```diff
@@ -655,11 +677,15 @@ require('lazy').setup({
       -- Better Around/Inside textobjects
       --
       -- Examples:
-      --  - `af` - around function call
-      --  - `inq` - inside next quote
+      --  - ya)  - [Y]ank [A]round [)]parenthen
+      --  - yinq - [Y]ank [I]nside [N]ext [']quote
+      --  - ci'  - [C]hange [I]nside [']quote
```
Comment (echasnovski) [inline init.lua:682]: Maybe something like this?

```
      --  - va)  - [V]isually select [A]round [)]parentheses 
      --  - yiq  - [Y]ank [I]nside [Q]uote
      --  - cin} - [C]hange [I]nside [N]ext [}]curly
```

It looks like more coverage.

File: init.lua [hunk 7]
```diff
@@ -44,60 +84,477 @@ P.S. You can delete this when you're done too. It's your config now :)
...
+          -- Show the signature of the function you're currently completing.
+          map('<C-k>', vim.lsp.buf.signature_help, 'Signature Documentation')
+        end,
+      })
+
+      -- LSP servers and clients are able to communicate to each other what features they support.
+      --  By default, Neovim doesn't support everything that is in the LSP Specification.
+      --  When you add nvim-cmp, luasnip, etc. Neovim now has *more* capabilities.
+      --  So, we create new capabilties with nvim cmp, and then broadcast that to the servers.
+      local capabilities = require('cmp_nvim_lsp').default_capabilities(vim.lsp.protocol.make_client_capabilities())
```
Comment (VonHeikemen) [inline init.lua:474]: You don't need to pass an argument to cmp_nvim_lsp `.default_capabilities()`.

`vim.lsp.protocol.make_client_capabilities()` doesn't have any effect on the return result of `.default_capabilities()`.

**Comment by rahulsoibam** [comment]
Getting this error on opening `init.lua` file on first open, not the subsequent ones.
OS: macOS 13.6.4
Neovim: `NVIM v0.9.5 Build type: Release LuaJIT 2.1.1703358377`

```
Error detected while processing function <SNR>33_NetrwBrowseChgDir[197]..BufReadPost Autocommands for "*":
Error executing lua callback: ...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:24: Error executing lua: ...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:25: fu
nction <SNR>33_NetrwBrowseChgDir[197]..BufReadPost Autocommands for "*"..FileType Autocommands for "*": Vim(append):Error executing lua callback: ...im/0.9.5/share/nvim/runtime/lua/
vim/treesitter/query.lua:259: query: invalid structure at position 3154 for language lua
stack traceback:
        [C]: in function '_ts_parse_query'
        ...im/0.9.5/share/nvim/runtime/lua/vim/treesitter/query.lua:259: in function 'get'
        ...5/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:114: in function 'new'
        ...r/neovim/0.9.5/share/nvim/runtime/lua/vim/treesitter.lua:61: in function '_create_parser'
        ...r/neovim/0.9.5/share/nvim/runtime/lua/vim/treesitter.lua:131: in function 'get_parser'
        ...r/neovim/0.9.5/share/nvim/runtime/lua/vim/treesitter.lua:460: in function 'start'
        ...m/lazy/nvim-treesitter/lua/nvim-treesitter/highlight.lua:20: in function 'attach'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:509: in function 'attach_module'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:532: in function 'reattach_module'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:133: in function <...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:132>
        [C]: in function 'nvim_cmd'
        ...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:25: in function <...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:24>
        [C]: in function 'nvim_buf_call'
        ...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:24: in function <...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_cmd'
        ...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:25: in function <...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:24>
        [C]: in function 'nvim_buf_call'
        ...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:24: in function <...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_buf_call'
        ...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:24: in function <...ocal/Cellar/neovim/0.9.5/share/nvim/runtime/filetype.lua:10>
```

**Comment by stseifer** [comment]
arch btw, 
NVIM v0.9.5
Build type: Release

Fails to start without error until 'help' parser removed from ensure install at line 775
`ensure_install = { 'c', 'lua', 'vim', 'vimdoc' }`

Not sure if help parser is needed

Error:
```
Error detected while processing /home/arch/.config/kickstart/init.lua:
Installation not possible: ...art/lazy/nvim-treesitter/lua/nvim-treesitter/install.lua:86: Parser not available for language "help"
See https://github.com/nvim-treesitter/nvim-treesitter/#adding-parsers on how to add a new parser!
```

**Review by feoh — APPROVED** [review]
Tested this for a few hours. Love this PR. More great work from @tjdevries !

File: init.lua [hunk 8]
```diff
@@ -106,27 +620,89 @@ require('lazy').setup({
...
-      'rafamadriz/friendly-snippets',
+      -- If you want to add a bunch of pre-configured snippets,
+      --    you can use this plugin to help you. It even has snippets
+      --    for various frameworks/libraries/etc. but you will have to
+      --    set up the ones that are useful for you.
+      -- 'rafamadriz/friendly-snippets',
     },
+    config = function()
+      -- See `:help cmp`
+      local cmp = require 'cmp'
```
Comment (sproclavism) [inline init.lua:645]: minor nitpick: there are a couple of function calls in the file without parentheses. Might be confusing to people that are not very familiar with lua syntax.

File: init.lua [hunk 9]
```diff
@@ -3,101 +3,615 @@
...
+  -- with the first argument being the link and the following
+  -- keys can be used to configure plugin behavior/loading/etc.
+  --
+  -- Use `opts = {}` to force a plugin to be loaded.
+  --
+  --  This is equivalent to:
+  --    require('Comment').setup({})
+
+  -- "gc" to comment visual regions/lines
+  { 'numToStr/Comment.nvim', opts = {} },
```
Comment (KaspervdHeijden) [inline init.lua:230]: Could also provide an event for lazy laouding:
```
event = { 'BufReadPost', 'BufNewFile' },
```

File: init.lua [hunk 10]
```diff
@@ -3,101 +3,615 @@
...
+      vim.keymap.set('n', '<leader>sn', function()
+        builtin.find_files {
+          cwd = vim.fn.stdpath 'config',
+        }
+      end, { desc = '[S]earch [N]eovim files' })
+    end,
+  },
+
+  { -- LSP Configuration & Plugins
     'neovim/nvim-lspconfig',
```
Comment (KaspervdHeijden) [inline init.lua:406]: Perhaps load this one lazy, too?
```
event  = 'VeryLazy',
```

File: init.lua [hunk 11]
```diff
@@ -137,144 +713,90 @@ require('lazy').setup({
...
-  {
-    -- Add indentation guides even on blank lines
-    'lukas-reineke/indent-blankline.nvim',
-    -- Enable `lukas-reineke/indent-blankline.nvim`
-    -- See `:help ibl`
-    main = 'ibl',
-    opts = {},
-  },
+  -- Highlight todo, notes, etc in comments
+  { 'folke/todo-comments.nvim', dependencies = { 'nvim-lua/plenary.nvim' }, opts = { signs = false } },
```
Comment (KaspervdHeijden) [inline init.lua:737]: For this one perhaps consider:
```
event  = { 'BufReadPost', 'BufNewFile' },
```

File: init.lua [hunk 12]
```diff
@@ -137,144 +713,90 @@ require('lazy').setup({
...
 
-  {
-    -- Highlight, edit, and navigate code
+  { -- Highlight, edit, and navigate code
     'nvim-treesitter/nvim-treesitter',
-    dependencies = {
-      'nvim-treesitter/nvim-treesitter-textobjects',
-    },
     build = ':TSUpdate',
+    config = function()
```
Comment (KaspervdHeijden) [inline init.lua:770]: Maybe?
```
event  = 'VeryLazy',
```

File: init.lua [hunk 13]
```diff
@@ -137,144 +713,90 @@ require('lazy').setup({
...
+      -- [[ Configure Treesitter ]] See `:help nvim-treesitter`
+
+      ---@diagnostic disable-next-line: missing-fields
+      require('nvim-treesitter.configs').setup {
+        ensure_installed = { 'c', 'lua', 'vim', 'vimdoc' },
```
Comment (MAHcodes) [inline init.lua:775]: I've had an issue while opening the README.md file due to no bash parser for treesitter, it might be frustrating for beginners to fix it, so if you can add bash to ensure_installed would be helpful.

my nvim version is:
```
NVIM v0.9.4
Build type: RelWithDebInfo
LuaJIT 2.1.1692716794
```

and here is the error:
```
Error detected while processing BufReadPost Autocommands for "*":
Error executing lua callback: /usr/local/share/nvim/runtime/filetype.lua:24: Error executing lua: /usr/local/share/nvim/runtime/filetype.lua:25: BufReadPost Autocommands for "*"..FileType Autocommands for "*": Vim(append):Error executing lu
a callback: /usr/local/share/nvim/runtime/lua/vim/treesitter/query.lua:259: query: invalid node type at position 196 for language bash
stack traceback:
        [C]: in function '_ts_parse_query'
        /usr/local/share/nvim/runtime/lua/vim/treesitter/query.lua:259: in function 'get'
        ...l/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:114: in function 'new'
        ...l/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:365: in function 'add_child'
        ...l/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:294: in function 'parse'
        /usr/local/share/nvim/runtime/lua/vim/treesitter.lua:89: in function '_create_parser'
        /usr/local/share/nvim/runtime/lua/vim/treesitter.lua:131: in function 'get_parser'
        /usr/local/share/nvim/runtime/lua/vim/treesitter.lua:459: in function 'start'
        ...m/lazy/nvim-treesitter/lua/nvim-treesitter/highlight.lua:20: in function 'attach'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:509: in function 'attach_module'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:532: in function 'reattach_module'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:133: in function <...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:132>
        [C]: in function 'nvim_cmd'
        /usr/local/share/nvim/runtime/filetype.lua:25: in function </usr/local/share/nvim/runtime/filetype.lua:24>
        [C]: in function 'nvim_buf_call'
        /usr/local/share/nvim/runtime/filetype.lua:24: in function </usr/local/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_cmd'
        /usr/local/share/nvim/runtime/filetype.lua:25: in function </usr/local/share/nvim/runtime/filetype.lua:24>
        [C]: in function 'nvim_buf_call'
        /usr/local/share/nvim/runtime/filetype.lua:24: in function </usr/local/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_buf_call'
        /usr/local/share/nvim/runtime/filetype.lua:24: in function </usr/local/share/nvim/runtime/filetype.lua:10>
Press ENTER or type command to continue
```

Comment (tjdevries) [reply init.lua:774]: good call, added

File: init.lua [hunk 14]
```diff
@@ -3,101 +3,615 @@
...
+          -- When you move your cursor, the highlights will be cleared (the second autocommand).
+          vim.api.nvim_create_autocmd({ 'CursorHold', 'CursorHoldI' }, {
+            buffer = event.buf,
+            callback = vim.lsp.buf.document_highlight,
+          })
+
+          vim.api.nvim_create_autocmd({ 'CursorMoved', 'CursorMovedI' }, {
+            buffer = event.buf,
+            callback = vim.lsp.buf.clear_references,
+          })
```
[the remark this answers is not in the document]
Comment (tjdevries) [reply init.lua]: yup, added, ty

File: init.lua (hunk 8 above)
Comment (tjdevries) [reply init.lua:644]: it's explained in the "learnxinyminutes" for lua, so I think it should be OK.

**Comment by tjdevries** [comment]
don't think i will add more lazy loading stuff -- not super worried about it. The ones that remain are fast enough or i want to make sure they load at the start so they know to install items

**Comment by tjdevries** [comment]
only remaining thing is that maybe i should add a checkhealth for kickstart.... so people could run that to check their environment and stuff. would work if they do the git clone method, but won't work if they do the copy-paste....

File: init.lua [hunk 15]
```diff
@@ -3,101 +3,614 @@
...
+        lua = { 'stylua' },
+        -- Conform can also run multiple formatters sequentially
+        -- python = { "isort", "black" },
+        --
+        -- You can use a sub-list to tell conform to run *until* a formatter
+        -- is found.
+        -- javascript = { { "prettierd", "prettier" } },
+      },
+    },
+  },
```
Comment (nhxnl) [inline init.lua:255]: Hey, it would be easier to read if this conform config is placed below the LSP/Mason config no?

**Review by rubiin — APPROVED** [review]

File: init.lua [hunk 16]
```diff
@@ -3,101 +3,614 @@
...
+          -- WARN: This is not Goto Definition, this is Goto Declaration.
+          --  For example, in C this would take you to the header
+          map('gD', vim.lsp.buf.declaration, '[G]oto [D]eclaration')
+
+          -- The following two autocommands are used to highlight references of the
+          -- word under your cursor when your cursor rests there for a little while.
+          --    See `:help CursorHold` for information about when this is executed
+          --
+          -- When you move your cursor, the highlights will be cleared (the second autocommand).
+          local client = vim.lsp.get_client_by_id(event.data.client_id)
```
Comment (KaspervdHeijden) [inline init.lua:507]: For what it's worth; there may be multiple clients attached, each may implement `documentHighlightProvider`.
Perhaps do something like this instead:

```lua

local added = false
for _, client in ipairs(vim.lsp.get_active_clients()) do
    if not added and client.server_capabilities.documentHighlightProvider then
        vim.api.nvim_create_autocmd({ 'CursorHold', 'CursorHoldI' }, {
            callback = vim.lsp.buf.document_highlight,
            buffer   = event.buf,
        })
                                                                                     
        vim.api.nvim_create_autocmd({ 'CursorMoved', 'CursorMovedI' }, {
            callback = vim.lsp.buf.clear_references,
            buffer   = event.buf,
        })

        added = true
    end
end
```

File: init.lua [hunk 17]
```diff
@@ -3,101 +3,614 @@
...
+      --  - settings (table): Override the default settings passed when initializing the server.
+      --        For example, to see the options for `lua_ls`, you could go to: https://luals.github.io/wiki/settings/
+      local servers = {
+        -- clangd = {},
+        -- gopls = {},
+        -- pyright = {},
+        -- rust_analyzer = {},
+        -- ... etc. See `:help lspconfig-all` for a list of all the pre-configured LSPs
+        --
+        -- If you use something like typescript, where the tooling is as bad as the language,
```
Comment (sodoku) [inline init.lua:545]: > as bad as the language

Can we please not judge other languages? I don't think this feels welcoming to typescript developers.

Comment (tjdevries) [reply init.lua:545]: Yes, I changed it for clarity, but not because I think jokes aren't funny

**Comment by feoh** [comment]
I know where he's coming from because I've thought about jumping off the kickstart, bandwagon myself, and doing a modular configuration of my own but the truth is the whole point of this project is to provide a monolithic starting point for new users, and so modularizing just isn't in the cards as I see it.

**Comment by destinio** [comment]
-1 for modular to start. I learned a lot by modularizing the base code myself :) 

Also, the ability to just quickly `curl` the raw init file is just too nice!!

File: init.lua (hunk 16 above)
Comment (tjdevries) [reply init.lua:507]: I think it's OK - the call is really cheap and it's fine to have this way. I agree something like this would probably be better, but not worth the confusion for users.

File: init.lua (hunk 17 above)
Comment (sodoku) [reply init.lua]: Thank you so much for making the change! I am all for jokes, I make jokes myself all the time :)

I just wanted to make sure that people who might not get this as a joke don't feel discouraged from using this amazing project as everything else is very beginner friendly.

**Comment by ro-i** [comment]
@tjdevries sorry for digging out this old pr, but I'd be curious whether there was a specific reason to change the default theme from onedark to tokyonight :)
